### PR TITLE
fix date logic for resetting users' pages printed

### LINKED ISSUE #1323: Correct date logic for resetting users' pages printed
Our current logic for resetting a user's available pages to print are incorrect. The logic that handles if we should reset a page can be found in https://github.com/SCE-Development/Clark/blob/af3a4be151d9697262ad04974e6502df4db58d65/api/main_endpoints/routes/Auth.js#L36

Below explains how the logic should work, but first the game plan

## Game Plan
1. move `checkIfPageCountResets` to https://github.com/SCE-Development/Clark/blob/af3a4be151d9697262ad04974e6502df4db58d65/api/main_endpoints/util/userHelpers.js
1. import the `checkIfPageCountResets`, now moved, in Auth.js
1. Create a new empty `describe` function in https://github.com/SCE-Development/Clark/blob/af3a4be151d9697262ad04974e6502df4db58d65/test/api/Auth.js#L67
this will be a unit test to test the function `checkIfPageCountResets`
1. run the empty `describe` function with `npm run api-test`
1. after you get to the above discuss with Evan the logic for below, and adding it to the function

## Logic

Below explains the logic we want for when we want to reset:

- if users last login was >= week ago, reset
- if users last login was under a week ago:
    - if users last login was on a sunday, do not reset, since we reset on sunday
    - if there was a sunday between the last login and now, reset

### As a table
for dates referenced, imagine we have a two week calendar like below:
image

| Last Login         | Login Time                  | Reset pages?  | Why?   |
| :----------------: | :------:                    | :----:        | :----  |
| Sunday the 1st     |   Sunday the 1st, same day  | Yes           | Because we just reset pages on sunday   |
| Sunday the 1st     |   Monday the 2nd,           | No            | Because we just reset pages earlier that day           |
| Sunday the 24th    |   Sunday the 1st            | Yes           | Because the difference between logins is >= 1 week           |
| Monday the 25th    |   Monday the 2nd            | Yes           | Because the difference between logins is >= 1 week           |
| Friday the 25th    |   Sunday the 1st            | Yes           | Because between the last login and now, there exist a sunday           |
| Friday the 25th    |   Thursday the 5th          | Yes           | Because between the last login and now, there exist a sunday           |

### PR DESCRIPTION
**Pull request by steeevin88**
removed old resetting logic
added proper logic to handle resetting a users' pages printed on Sundays (when user's last login was 7+ days ago or a Sunday passed in between logins)

added function to mock day, month, and year
added tests to check reset logic

resolves #1323

### CODE REVIEW

File: api/main_endpoints/util/userHelpers.js [hunk 1]
```diff
@@ -147,10 +147,38 @@ function hashPassword(password) {
...
+
+  let oldDate = new Date(lastLogin.getTime());
+
+  // this returns "right now" when called
+  // to unit test, we mock when "right now" is
+  //
+  // by mocking, we can have `new Date` return
+  // tomorrow, last week etc
+  const now = new Date();
+  const week = 7 * 24 * 60 * 60 * 1000; // a week in milliseconds
```
Comment (evanugarte) [inline api/main_endpoints/util/userHelpers.js:168]:
```suggestion
  const oneWeekInMilliseconds = 7 * 24 * 60 * 60 * 1000; // a week in milliseconds
```

File: test/api/Auth.js [hunk 2]
```diff
@@ -189,4 +192,44 @@ describe('Auth', () => {
     });
   });
 
+  describe('checkIfPageCountResets()', () => {
+    it('Should reset user\'s page count when last login was over 7 days ago',
```
Comment (evanugarte) [inline test/api/Auth.js:196]: let's skip using `\`s, we can rephrase the titles like
```suggestion
    it('Should reset page count when the last login was over 7 days ago',
```

File: test/api/Auth.js [hunk 3]
```diff
@@ -189,4 +192,44 @@ describe('Auth', () => {
...
+      async () => {
+        // mock current day to January 10th, 2023 (Tuesday)
+        const mockCurrentDate = mockDayMonthAndYear(10, 0, 2023);
+        // mock last login to January 1st, 2023 (Sunday)
+        const mockLastLogin = new Date(2023, 0, 1);
+        const result = checkIfPageCountResets(mockLastLogin);
+        expect(result).to.be.true;
+      });
+
+    it('Should reset user\'s page count when there\s a Sunday between last login and now',
```
Comment (evanugarte) [inline test/api/Auth.js:206]: another example rename, the other ones should do the same thing

```suggestion
    it('Should reset page count when there is a Sunday between last login and now',
```

File: api/main_endpoints/util/userHelpers.js [hunk 4]
```diff
@@ -147,10 +147,39 @@ function hashPassword(password) {
...
+  const now = new Date();
+  const oneWeekInMilliseconds = 7 * 24 * 60 * 60 * 1000; // a week in milliseconds
+  // reset if users last login was >= week ago OR there was a sunday between the last login and now
+  if (now.getTime() - oldDate.getTime() >= oneWeekInMilliseconds
+  || oldDate.getDay() > now.getDay()) {
+    return true;
+  }
+
+  return false;
+}
```
Comment (evanugarte) [inline api/main_endpoints/util/userHelpers.js:176]: slightly cleaner
```suggestion
  const lastLoginWasOverOneWeekAgo = now.getTime() - oldDate.getTime() >= oneWeekInMilliseconds;
  const aSundayHasPassedSinceLastLogin = oldDate.getDay() > now.getDay();
  return lastLoginWasOverOneWeekAgo || oneWeekInMilliseconds;
}
```

File: api/main_endpoints/util/userHelpers.js [hunk 5]
```diff
@@ -147,10 +147,39 @@ function hashPassword(password) {
...
+
+  let oldDate = new Date(lastLogin.getTime());
+
+  // this returns "right now" when called
+  // to unit test, we mock when "right now" is
+  //
+  // by mocking, we can have `new Date` return
+  // tomorrow, last week etc
+  const now = new Date();
+  const oneWeekInMilliseconds = 7 * 24 * 60 * 60 * 1000; // a week in milliseconds
```
Comment (evanugarte) [inline api/main_endpoints/util/userHelpers.js:168]: we can remove the comment the variable name explains it
```suggestion
  const oneWeekInMilliseconds = 7 * 24 * 60 * 60 * 1000;
```

**Review by evanugarte — APPROVED** [review]

**Review by joshua-demo — APPROVED** [review]
lgtm

**Pull request merged**